src/compactor.rs

```diff
@@ -18,11 +18,30 @@ use crate::error::SlateDBError;
 use crate::manifest_store::{FenceableManifest, ManifestStore, StoredManifest};
 use crate::metrics::DbStats;
 use crate::tablestore::TableStore;
+use crate::types::{KeyValueDeletable, ValueDeletable};
 
 pub trait CompactionScheduler {
     fn maybe_schedule_compaction(&self, state: &CompactorState) -> Vec<Compaction>;
 }
 
+pub trait CompactionFilter: Send + Sync {
+    fn filter(&self, kv: &KeyValueDeletable) -> FilterResult;
+}
+
+pub enum FilterResult {
```

Comment:
I am wondering if we can do this without adding a new `FilterResult` enum type. `KeyValueDeletable` already has the concept of delete (tombstone). I guess there are a few operations we want to encapsulate:

1. Leave as-is (whether the `ValueDeletable` is a `ValueDeletable::Tombstone` or `ValueDeletable::value()`
2. Drop the `KeyValueDeletable` (don't write a tombstone at all)
3. Delete the `KeyValueDeletable` (by returning a KeyValueDeletable` with a tombstone for the value)
4. Modify the `KeyValueDeletable`. This could potentially include changing the key, the value, or even the metadata (e.g. timestamp if we support that).

If we had:

```rust
pub trait CompactionFilter: Send + Sync {
    fn filter(&self, kv: &KeyValueDeletable) -> Option<KeyValueDeletable>;
}
```

I think we could address 1-4 as:

1. Return `kv`.
2. Return None.
3. Return a new `KeyValueDeletable` with `kv`'s key and a `ValueDeletable::Tombstone` value.
4. Return a new `KeyValueDeletable` with key and value set as required by filter logic.

Reply:
I added it because I don't think we should support changing the key -- that opens the door to some pretty gnarly edge cases I don't want to think about (RocksDB doesn't support changing the key). If you change the key, the new key could (for example) belong to a different SST within the Sorted Run that you've already compacted (so that would break the invariant there). It could also cause an implicit overwrite of another key, which means we'd have to consider what that means for the TTL/insertion timestamp etc... For now it's much simpler not to support key modification.

The other benefit of adding this layer of indirection is that it lets us evolve the API independently of the `KeyValueDeletable` API. An example here would be if we wanted to add a "soft filter" result or something like that for when you chain filters (imagine we have three filters, and one will only insert a tombstone if none of the other filters modified the value). That would be challenging to encode directly into the `ValueDeletable`

Reply:
Another extension (which I would argue against, but goes to show the flexibility here) is to introduce a `FlatMap` filter result that allows you to actually map one key -> many during compaction.

Reply:
Good point on the key. I hadn't considered that.

One nit: could we call it `CompactionFilterResult` then? Filter is an overloaded term that also means bloom filter all over the codebase.

Reply:
🫡 can do! 

Also as part of the TTL RFC process I'm thinking we might want to leave the public API for compaction filters for another time and just implement one for TTL that will run if the SST has TTL enabled (this will make more sense once the RFC is out there). I'll just design it with the intention of one day exposing compaction filters.

Reply:
> we might want to leave the public API for compaction filters for another time and just implement one for TTL that will run if the SST has TTL enabled

💯 this is what I just concluded in my comments on #225